src/parser/statement.js

```diff
@@ -1009,21 +1019,43 @@ export default class StatementParser extends ExpressionParser {
       return;
     }
 
+    const isPrivate = this.match(tt.hash);
+    if (isPrivate) {
+      this.expectOnePlugin(["classPrivateProperties", "classPrivateMethods"]);
```

Comment:
Instead of throwing here, you can wait until it is known if the node is a property or a method (to give a more specific error message).